### ISSUE TITLE
(sup.) forecasting interface design

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#18]
aka: FK and AG to finish their paper

**Comment by mloning** [comment]
Here are some ideas for the forecasting `task` object thanks to @sajaysurya. The main intuition was to keep it as simple as possible and similar to the TSC task interface, possibly allowing us to have a single task class rather than one for each use case. Going through @fkiraly's design above: 

> supervised - boolean (corresponds to (1) above)

may not be necessary, as it's identifiable from data/metadata (one or multiple rows)
 
> predictfrom - list of names of covariates/inputs/features (columns) to predict from
> (in prediction, features are used only until observation cut-off)

is identical to features in TSC task

> predicttargets - list of names of targets/outputs/labels (columns) to predict

is identical to target in TSC task

> usetargettoforecast - boolean. If "false", for forecasting, methods may not access the target time series, even if observations lie in the past. If "true" (default), the targets are used for forecasting and visible to the forecaster method.

not necessary as a separate argument, target column could simply be both in features and target

> obshorizon - dictionary with following fields
> obshorizon.type - one of "single" (default), "moving", "custom"

For classical forecasting, obshorizon could be handled with a time series splitter (similar to train_test_split, but for time indices), which supports fixed and moving splits and which accepts one or multiple fixed cut-off points or instructions for a moving cut-off point, possibly refactoring the sklearn's [TimeSeriesSplitter](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html) 

For the supervised forecasting case, the splitter would be a 2-dimensional splitter, splitting both temporally and across iid samples, e.g. 

`(train_rows, train_times), (test_rows, test_times) = SupForecastCrossValidator().split()`

For predictions at multiple cut-off points, the `orchestrator` could be used to automatically take care of the splitting, fitting and predicting. 

> predhorizon - dictionary with following fields
> predhorizon.type - one of "concrete" (default), "functional"
> predhorizon.anchor - one of "relative" (default), "absolute"

predhorizon has to be in task but could for example be passed together with target as a tuple to the task, i.e. `target=(target_name, predhorizon)`, rather than a separate argument.

**Comment by mloning** [comment]
* transformer interface for transformation compatible with multiple-sample (supervised) and single-sample forecasting setting
* in-sample forecasts
* forecast intervals
* update functionality

**Comment by fkiraly** [comment]
For reference, I'm uploading the notes I wrote in preparation to last week's discussion on @mloning 's architecture document
https://github.com/alan-turing-institute/sktime/wiki/Forecasting-API-proposal
[Forecasting design comments.docx](https://github.com/alan-turing-institute/sktime/files/3887577/Forecasting.design.comments.docx)

**Comment by mloning** [comment]
See #218 and corresponding [API design document](https://github.com/alan-turing-institute/sktime/wiki/Forecasting-API-proposal)
